File: packages/core/src/towers.ts
import { CellTower } from "./types";

/**
 * Cell Tower Utilities
 *
 * Functions for working with cell tower data and calculating
 * distances for spider-leg visualizations.
 */

/**
 * Calculate the distance between two points using Haversine formula
 * @returns Distance in kilometers
 */
export function calculateDistance(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number
): number {
  const R = 6371; // Earth's radius in kilometers
  const dLat = toRadians(lat2 - lat1);
  const dLng = toRadians(lng2 - lng1);

  const a =
    Math.sin(dLat / 2) * Math.sin(dLat / 2) +
    Math.cos(toRadians(lat1)) *
      Math.cos(toRadians(lat2)) *
      Math.sin(dLng / 2) *
      Math.sin(dLng / 2);

  const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));
  return R * c;
}

function toRadians(degrees: number): number {
  return degrees * (Math.PI / 180);
}

/**
 * Find the N nearest cell towers to a given point
 */
export function findNearestTowers(
  lat: number,
  lng: number,
  towers: CellTower[],
  count: number = 3
): CellTower[] {
  if (towers.length === 0) return [];

  const towersWithDistance = towers.map((tower) => ({
    tower,
    distance: calculateDistance(lat, lng, tower.lat, tower.lng),
  }));

  // Sort by distance and take the nearest N
  towersWithDistance.sort((a, b) => a.distance - b.distance);

  return towersWithDistance.slice(0, count).map((t) => t.tower);
}

/**
 * Generate spider-leg polyline coordinates
 * Returns an array of [lat, lng] pairs for drawing lines from a point to towers
 */
export function generateSpiderLegs(
  centerLat: number,
  centerLng: number,
  towers: CellTower[]
): Array<[[number, number], [number, number]]> {
  return towers.map((tower) => [
    [centerLat, centerLng],
    [tower.lat, tower.lng],
  ]);
}

/**
 * Get the icon type for a tower based on its type
 */
export function getTowerIcon(type: "4G" | "5G"): string {
  // Lucide icon names
  return type === "5G" ? "zap" : "antenna";
}

/**
 * Format tower info for display
 */
export function formatTowerInfo(tower: CellTower): string {
  return `${tower.type} Tower (Cell ID: ${tower.cellId})`;
}

/**
 * Filter towers within a bounding box
 */
export function filterTowersInBounds(
  towers: CellTower[],
  north: number,
  south: number,
  east: number,
  west: number
): CellTower[] {
  return towers.filter((tower) => {
    return (
      tower.lat >= south &&
      tower.lat <= north &&
      tower.lng >= west &&
      tower.lng <= east
    );
  });
}

/**
 * Group towers by type for statistics
 */
export function groupTowersByType(
  towers: CellTower[]
): Record<"4G" | "5G", number> {
  const counts = { "4G": 0, "5G": 0 };

  for (const tower of towers) {
    counts[tower.type]++;
  }

  return counts;
}

/**
 * Calculate average distance from a point to all provided towers
 */
export function calculateAverageTowerDistance(
  lat: number,
  lng: number,
  towers: CellTower[]
): number {
  if (towers.length === 0) return 0;

  const totalDistance = towers.reduce(
    (sum, tower) => sum + calculateDistance(lat, lng, tower.lat, tower.lng),
    0
  );

  return totalDistance / towers.length;
}

File: apps/mobile/lib/domain/towers.dart
import 'dart:math' as math;

import 'entities/cell_tower.dart';

/// Cell Tower Utilities
///
/// Functions for working with cell tower data and calculating distances for
/// spider-leg visualizations.

/// Haversine distance between two points, in kilometers.
double calculateDistance(double lat1, double lng1, double lat2, double lng2) {
  const r = 6371.0; // Earth's radius in kilometers
  final dLat = _toRadians(lat2 - lat1);
  final dLng = _toRadians(lng2 - lng1);

  final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
      math.cos(_toRadians(lat1)) *
          math.cos(_toRadians(lat2)) *
          math.sin(dLng / 2) *
          math.sin(dLng / 2);

  final c = 2 * math.atan2(math.sqrt(a), math.sqrt(1 - a));
  return r * c;
}

double _toRadians(double degrees) => degrees * (math.pi / 180);

List<CellTower> findNearestTowers(
  double lat,
  double lng,
  List<CellTower> towers, [
  int count = 3,
]) {
  if (towers.isEmpty) return [];

  final towersWithDistance = [
    for (final tower in towers) (tower: tower, distance: calculateDistance(lat, lng, tower.lat, tower.lng)),
  ];

  towersWithDistance.sort((a, b) => a.distance.compareTo(b.distance));

  return [for (final t in towersWithDistance.take(count)) t.tower];
}

/// Spider-leg segments: pairs of (lat, lng) points, one per tower, from the
/// center point to that tower.
List<((double, double), (double, double))> generateSpiderLegs(
  double centerLat,
  double centerLng,
  List<CellTower> towers,
) {
  return [
    for (final tower in towers) ((centerLat, centerLng), (tower.lat, tower.lng)),
  ];
}

String getTowerIcon(CellTowerType type) {
  // Lucide icon names
  return type == CellTowerType.fiveG ? 'zap' : 'antenna';
}

String towerTypeLabel(CellTowerType type) => type == CellTowerType.fiveG ? '5G' : '4G';

String formatTowerInfo(CellTower tower) {
  return '${towerTypeLabel(tower.type)} Tower (Cell ID: ${tower.cellId})';
}

List<CellTower> filterTowersInBounds(
  List<CellTower> towers,
  double north,
  double south,
  double east,
  double west,
) {
  return towers
      .where((tower) => tower.lat >= south && tower.lat <= north && tower.lng >= west && tower.lng <= east)
      .toList();
}

Map<CellTowerType, int> groupTowersByType(List<CellTower> towers) {
  final counts = {CellTowerType.fourG: 0, CellTowerType.fiveG: 0};

  for (final tower in towers) {
    counts[tower.type] = counts[tower.type]! + 1;
  }

  return counts;
}

double calculateAverageTowerDistance(double lat, double lng, List<CellTower> towers) {
  if (towers.isEmpty) return 0;

  final totalDistance = towers.fold<double>(0, (sum, tower) => sum + calculateDistance(lat, lng, tower.lat, tower.lng));

  return totalDistance / towers.length;
}
